Better determination of runnable CPU count

ac000 opened the pull request:
```
This pull-request comprises five commits

- Configure: add a check for sched_getaffinity(2)

Adds a configure check for sched_getaffinity(2)

- Configure: add cpuset_getaffinity() support

Adds a configure check for cpuset_getaffinity(2)

- Configure: add a check for the CPU_COUNT() macro

Adds a configure check for CPU_COUNT()

This does detect CPU_COUNT() on FreeBSD 13.1+ twice due to the
introduction of sched.h which includes sys/cpuset.h so we end up with it
detected twice. This isn't really an issue due to the feature macros
being guarded against multiple definitions.

Fixing this cosmetic issue gets messy. I think really the right thing is
to move the sched_[gs]etaffinity() and CPU_COUNT() checks out of
auto/unix into auto/os/linux as AFAICT only Linux has those syscalls (a
part from FreeBSD with it's compatibility wrappers, but we use the
native implementation there anyway).

Or we simply drop this check and assume CPU_COUNT() is there which is
true for any Linux/FreeBSD system in the last decade and more.

- Better available cpu count determination on Linux

Uses the above to better determine the actual number of runnable CPUs.

Uses the static allocation API, this is simpler than the dynamic API,
which has tricky edge cases around trying to determine the number of
CPUs that need to be covered, and supports up to 1024 CPUs. If a system
has more, then we'll just fall back to using the original ngx_ncpu.

- Better available cpu count determination on FreeBSD

FreeBSD only has a static allocation API and while the userspace bits
support 1024 CPUs (like on Linux), however it seems the Kernel currently
limits this to 256 (MAX_CPU).

Why This?
---------

nginx currently uses 'sysconf(_SC_NPROCESSORS_ONLN)' (when available),
to determine the number of CPUs in the system to determine how many
worker process to create with 'worker_processes auto;'

Often this produces the correct result, nr worker processes == runnable
cpu count. However there are cases where this can go awry.

For example when cpuset(7)s are in use and a process has been limited to
a subset of CPUs.

E.g. a common case is docker with the '--cpuset-cpus=' argument set.

If the host has say 16 CPUs and docker has been allocated 4, then nginx
would still create 16 workers under docker. After this change it will
create 4.
```

ac000 committed:
Configure: add a check for sched_getaffinity(2)

This will help to better determine the number of worker processes to
create in certain situations on Linux.

Unlike sysconf(_SC_NPROCESSORS_ONLN) this takes into account per-process
cpu allowed masks as set by sched_setaffinity(2)/cpusets etc.

So while a system may have 64 on-line cpu's, nginx itself may be limited
to using just four of them in which case we should create just four
worker processes not 64 under 'worker_processes auto;'

ac000 committed:
Configure: add a check for sched_getaffinity(2).

This will help to better determine the number of worker processes to
create in certain situations.

Unlike sysconf(_SC_NPROCESSORS_ONLN) this takes into account per-process
cpu allowed masks as set by sched_setaffinity(2)/cpusets etc.

So while a system may have 64 on-line cpu's, nginx itself may be limited
to using just four of them in which case we should create just four
worker processes not 64 under 'worker_processes auto;'

sched_getaffinity(2) is available on at least Linux & FreeBSD.

ac000 committed:
Better available cpu count determination.

With 'worker_processes auto;' the number of worker processes to create
is generally determined at runtime via sysconf(_SC_NPROCESSORS_ONLN);

However this value does not take into account any restrictions that have
been put in place via mechanisms like cpuset(7)'s and/or
sched_setaffinity(2).

So for example while a system may have 18 "cpus" available as returned
via sysconf(_SC_NPROCESSORS_ONLN) the actual number of allowed to run on
cpus may be much lower.

E.g.

On a 18 core system, with 'worker_processes auto;' nginx will create 18
worker processes (+ the master). E.g.

  $ pgrep -c nginx
  19

If however nginx has been restricted to run on just four. We will still
create 18 worker processes. E.g.

  $ taskset -a -c 0-3 ./sbin/nginx -c conf/minimal.conf
  $ pgrep -c nginx
  19

This can happen for example with 'docker --cpuset-cpus=', where you may
end up with many more worker processes than runnable cpus.

This commit uses sched_getaffinity() (where available, at least Linux &
FreeBSD) to determine the number of actual available runnable cpus.

So in the above example we now get

  $ taskset -a -c 0-3 ./sbin/nginx -c conf/minimal.conf
  $ pgrep -c nginx
  5

Four worker processes plus the master.

Closes: https://github.com/nginx/nginx/issues/855

ac000 committed:
Better available cpu count determination on Linux

With 'worker_processes auto;' the number of worker processes to create
is generally determined at runtime via sysconf(_SC_NPROCESSORS_ONLN).

However this value does not take into account any restrictions that have
been put in place via mechanisms like cpuset(7)'s and/or
sched_setaffinity(2).

So for example while a system may have 18 "cpus" available as returned
via sysconf(_SC_NPROCESSORS_ONLN) the actual number of allowed to run on
cpus may be much lower.

E.g.

On a 18 core system, with 'worker_processes auto;' nginx will create 18
worker processes (+ the master). E.g.

  $ pgrep -c nginx
  19

If however nginx has been restricted to run on just four. We will still
create 18 worker processes. E.g.

  $ taskset -a -c 0-3 ./sbin/nginx -c conf/minimal.conf
  $ pgrep -c nginx
  19

This can happen for example with 'docker --cpuset-cpus=', where you may
end up with many more worker processes than runnable cpus.

This commit uses the Linux sched_getaffinity() system-call to determine
the number of actual available runnable cpus.

So in the above example we now get

  $ taskset -a -c 0-3 ./sbin/nginx -c conf/minimal.conf
  $ pgrep -c nginx
  5

Four worker processes plus the master.

Closes: https://github.com/nginx/nginx/issues/855

ac000 commented:
REbased with master

```
$ git range-diff 2b8874d07...1c59ff16d
 -:  --------- >  1:  eb5ebbbed QUIC: fixed ssl_reject_handshake error handling.
 -:  --------- >  2:  417c87b78 Updated link to xslscript.
 -:  --------- >  3:  bc71625dc Fixed inaccurate index directive error report.
 -:  --------- >  4:  0373fe5d9 SNI: using the ClientHello callback.
 -:  --------- >  5:  7f9ced0ce SNI: support for early ClientHello callback with BoringSSL.
 -:  --------- >  6:  4c9ae11df QUIC: localized OpenSSL headers used for QUIC protection.
 -:  --------- >  7:  af436c58c QUIC: a new macro to differentiate BoringSSL specific EVP API.
 -:  --------- >  8:  93ff1ee12 SSL: AWS-LC compatibility.
 -:  --------- >  9:  8255bd29a Upstream: overflow detection in Cache-Control delta-seconds.
 -:  --------- > 10:  6f81314a0 Mail: xtext encoding (RFC 3461) in XCLIENT LOGIN.
 -:  --------- > 11:  c2a266fa7 Added F5 CLA workflow.
 -:  --------- > 12:  a144d828c SSL: fixed "key values mismatch" with object cache inheritance.
 -:  --------- > 13:  9d5cdc616 Fixed a typo.
 -:  --------- > 14:  7f71abdd1 nginx-1.29.2-RELEASE
 -:  --------- > 15:  3494f25c3 Version bump.
 -:  --------- > 16:  f5a989cda Updated OpenSSL and PCRE used for win32 builds.
 -:  --------- > 17:  25b03d650 SSL: disabled using certificate compression with OCSP stapling.
 -:  --------- > 18:  78d1ab5a2 SSL: support for compressed server certificates with BoringSSL.
 -:  --------- > 19:  c8c7beb96 Added $request_port and $is_request_port variables.
 -:  --------- > 20:  42ca3a457 CONNECT method support for HTTP/1.1.
 -:  --------- > 21:  364a94ece Upstream: reset local address in case of error.
 -:  --------- > 22:  71f8eb52b SSL: $ssl_sigalg, $ssl_client_sigalg.
 -:  --------- > 23:  ac72ca60c Geo: the "volatile" parameter.
 -:  --------- > 24:  c93a0c48a Headers filter: inheritance control for add_header and add_trailer.
 -:  --------- > 25:  ce30a1cb0 OCSP: fixed invalid type for the 'ssl_ocsp' directive.
 -:  --------- > 26:  f04e2b7f6 Fixed compilation warnings on Windows after c93a0c48af87.
 -:  --------- > 27:  65c0b2e77 Modules compatibility: increased compat section size.
 -:  --------- > 28:  fd0848bdd nginx-1.29.3-RELEASE
 -:  --------- > 29:  99312be10 Configure: ensure we get the "built by ..." line in nginx -V.
 -:  --------- > 30:  a386df6b2 Version bump.
 -:  --------- > 31:  ac99f2808 Configure: MSVC compatibility with PCRE2 10.47.
 -:  --------- > 32:  38a701d88 SSL: ngx_ssl_set_client_hello_callback() error handling.
 -:  --------- > 33:  9d04b6630 SSL: fixed build with BoringSSL, broken by 38a701d88.
 -:  --------- > 34:  6ed118841 HTTP/2: extended guard for NULL buffer and zero length.
 -:  --------- > 35:  511abb19e Improved host header validation.
 -:  --------- > 36:  6446f9910 Changed interface of ngx_http_validate_host().
 -:  --------- > 37:  bcb41c919 Proxy: fixed segfault in URI change.
 1:  fdba0e119 = 38:  453b5b5ca Configure: add a check for sched_getaffinity(2).
 2:  2b8874d07 = 39:  1c59ff16d Better available cpu count determination.
```

pcty-mbrandeis commented:
@ac000 & @pluknet  - What is the latest on this PR? I see it was opened a few months ago (5 as of this message). Is there anything blocking merging this?

pluknet commented:
@ac000 

This would need a similar change for cpuset_getaffinity().
(I have an old patch from 2015 to share.)

Also, it is not clear how this respects this comment:
https://trac.nginx.org/nginx/ticket/1151#comment:1
(Personally, I don't see/recall there a problem. You might have more details after a similar contribution to Unit.)

ac000 commented:
Hi @pluknet

> This would need a similar change for cpuset\_getaffinity()

Heh, I saw a sched\_getaffinity(2) man page on the FreeBSD website and
thought great, they have the same interface. Now I realise it's just
them showing Linux man-pages!

> Also, it is not clear how this respects this comment:                       
> https://trac.nginx.org/nginx/ticket/1151#comment:1

I'm not sure I really understand the concern.

In either case the cpu mask returned by sched\_getaffinity(2) is what's
available.

> (Personally, I don't see/recall there a problem. You might have more
> details after a similar contribution to Unit.)

In Unit's case this was originally brought up due to someone seeing some
lock contention due to Unit creating nr router threads == nr cpus. E.g.
a 64 core system with unit running under docker limited to 4 cores,
would still create 64 router threads.

Running Unit under docker was pretty common so in either case it seemed
wise to be a bit more accurate in how many router threads we were
creating.

In nginx's case I'm not sure what the _actual_ issues are other than
the original report

> Currently, if one uses worker\_processes auto; , nginx uses                 
> sysconf(\_SC\_NPROCESSORS\_ONLN) to detect the number of available
> processes. 
                                                                              
> In environments, such as Kubernetes or OpenShift, this causes issues,
> as it returns the cores available to the system and not the actual
> services available to the service. 

So aside from the expectation that `worker_processes auto` creates nr
processes == available runnable cpus...

ac000 committed:
Configure: add cpuset_getaffinity() support

This adds FreeBSD cpuset_getaffinity() support alongside the existing
cpuset_setaffinity().

This will be used to better determine the number of actual runnable
CPUs on FreeBSD.

ac000 committed:
Better available cpu count determination on FreeBSD

For the reasons outlined in the previous commit.

We prefer the native FreeBSD cpuset_getaffinity() system-call over the
Linux sched_getaffinity() compatibility wrapper added in 13.1.

Closes: https://github.com/nginx/nginx/issues/855

ac000 commented on src/os/unix/ngx_posix_init.c:
```diff
@@ -64,6 +64,29 @@ ngx_os_init(ngx_log_t *log)
+
+        err = cpuset_getaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, -1,
+                                 sizeof(mask), &mask);
+        if (err == 0) {
+            ngx_ncpu = CPU_COUNT(&mask);
+        }
```
I think it's better just to add a check for the CPU_COUNT() macro than do all that...

ac000 committed:
Configure: add a check for the CPU_COUNT() macro

This is used when working with cpuset bitmasks to return the number of
CPUs in a set.

ac000 commented on src/os/unix/ngx_posix_init.c:
```diff
@@ -64,6 +64,29 @@ ngx_os_init(ngx_log_t *log)
+        cpu_set_t  mask;
+
+        err = sched_getaffinity(0, sizeof(mask), &mask);
+        if (err == 0) {
+            ngx_ncpu = CPU_COUNT(&mask);
+        }
```
We have a check for CPU_COUNT()

ac000 commented on src/os/unix/ngx_posix_init.c:
```diff
@@ -62,6 +62,33 @@ ngx_os_init(ngx_log_t *log)
+
+        err = cpuset_getaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, -1,
+                                 sizeof(mask), &mask);
+        if (err == 0) {
+            ngx_ncpu = CPU_COUNT(&mask);
+        }
```
I think this is most unlikely and we have the if (ngx_cpu < 1) check as a safety net.

ac000 commented on src/os/unix/ngx_posix_init.c:
```diff
@@ -62,6 +62,33 @@ ngx_os_init(ngx_log_t *log)
+        cpu_set_t  mask;
+
+        err = sched_getaffinity(0, sizeof(mask), &mask);
+        if (err == 0) {
+            ngx_ncpu = CPU_COUNT(&mask);
+        }
```
Ditto.

ac000 commented on auto/unix:
```diff
@@ -292,6 +292,28 @@ ngx_feature_test="cpu_set_t mask;
+ngx_feature="CPU_COUNT()"
+ngx_feature_name="NGX_HAVE_CPU_COUNT"
+ngx_feature_run=no
+ngx_feature_incs="#include <sched.h>"
+ngx_feature_path=
+ngx_feature_libs=
+ngx_feature_test="cpu_set_t set;
+                  CPU_COUNT(&set);"
```
On FreeBSD cpu_set_t is typdef'd to cpuset_t. But that wasn't added till 13.1. So yeah, should probably update the check so it will work on older FreeBSDs...

ac000 replied:
Fixed a different way.
